### PR TITLE
[LLC] Add test server tests

### PR DESCRIPTION
**Pull request by chamons**
Fixes: https://github.com/Azure/autorest.csharp/issues/1115

The 5 tests were copied from the data plane tests and ported. Here are the diffs between data plane and LLC versions:

body-string -> https://gist.github.com/chamons/ce649e1684a09583a18cb578e8e160ac
url -> https://gist.github.com/chamons/7a4e169d13bec96eceba53a2317d7d63
url-query -> https://gist.github.com/chamons/03cb5df6b1a4dd1e2b40145d2280d751
url-path-items -> https://gist.github.com/chamons/ab59f4096fcf5805c571a1342c347bdf
body-complex -> Too many deletions for diff to be useful 

This is a diff from tldr-patch of the non-generated files from this PR with the tests themselves removed that may be helpful:

https://gist.github.com/chamons/8c36c0703fe67d8adce27e15a2188461

### CODE REVIEW

File: test/AutoRest.TestServerLowLevel.Tests/url.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        [Test]
+        public Task UrlPathsFloatNegative() => TestStatus(async (host, key) => await new PathsClient(key, host).FloatScientificNegativeAsync());
+
+        [Test]
+        public Task UrlPathsDoubleNegative() => TestStatus(async (host, key) => await new PathsClient(key, host).DoubleDecimalNegativeAsync());
+
+        [Test]
+        public Task UrlPathsDoublePositive() => TestStatus(async (host, key) => await new PathsClient(key, host).DoubleDecimalPositiveAsync());
+
+        public override IEnumerable<string> AdditionalKnownScenarios { get; } = new string[] {"UrlPathsBoolFalse",
```
Comment (pakrym) [inline test/AutoRest.TestServerLowLevel.Tests/url.cs:97]: I think we can delete this. We deleted the coverage tracking funcitonality.

File: test/AutoRest.TestServerLowLevel.Tests/body-complex.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+        [Test]
+        public Task PutComplexBasicValid() => TestStatus(async (host, key) =>
+        {
+            var data = new JsonData(new Dictionary<string, object> {
+                ["name"] = "abc",
+                ["id"] = 2,
+                ["color"] = "Magenta"
+            });
+            return await new BasicClient(key, host).PutValidAsync(RequestContent.Create(data));
```
Comment (pakrym) [inline test/AutoRest.TestServerLowLevel.Tests/body-complex.cs:42]: I think we don't need to pass the key every time. We can add a static Key on the TestServerLowLevelTestBase

File: src/AutoRest.CSharp/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -3,4 +3,5 @@
...
\ No newline at end of file
+[assembly:InternalsVisibleTo("AutoRest.TestServer.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100d15ddcb29688295338af4b7686603fe614abd555e09efba8fb88ee09e1f7b1ccaeed2e8f823fa9eef3fdd60217fc012ea67d2479751a0b8c087a4185541b851bd8b16f8d91b840e51b1cb0ba6fe647997e57429265e85ef62d565db50a69ae1647d54d7bd855e4db3d8a91510e5bcbd0edfbbecaa20a7bd9ae74593daa7b11b4")]
+[assembly:InternalsVisibleTo("AutoRest.TestServerLowLevel.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100d15ddcb29688295338af4b7686603fe614abd555e09efba8fb88ee09e1f7b1ccaeed2e8f823fa9eef3fdd60217fc012ea67d2479751a0b8c087a4185541b851bd8b16f8d91b840e51b1cb0ba6fe647997e57429265e85ef62d565db50a69ae1647d54d7bd855e4db3d8a91510e5bcbd0edfbbecaa20a7bd9ae74593daa7b11b4")]
```
Comment (pakrym) [inline src/AutoRest.CSharp/Properties/AssemblyInfo.cs:7]: Curious, what internals do we use?

**Review by pakrym — APPROVED** [review]
Sweet!

File: src/AutoRest.CSharp/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (chamons) [reply src/AutoRest.CSharp/Properties/AssemblyInfo.cs:7]: test\AutoRest.TestServer.Tests\Infrastructure\JsonAsserts.cs(11,65): error CS0122: 'IUtf8JsonSerializable' is inaccessible due to its protection level 
test\AutoRest.TestServer.Tests\Infrastructure\JsonAsserts.cs(25,52): error CS0122: 'IUtf8JsonSerializable' is inaccessible due to its protection level
test\AutoRest.TestServer.Tests\Infrastructure\TestServerTestBase.cs(23,25): error CS0122: 'ClientDiagnostics' is inaccessible due to its protection level

File: test/AutoRest.TestServerLowLevel.Tests/url.cs (hunk 1 above)
Comment (chamons) [reply test/AutoRest.TestServerLowLevel.Tests/url.cs:97]: Want me to nuke it from all tests or just my new ones?

Comment (pakrym) [reply test/AutoRest.TestServerLowLevel.Tests/url.cs:97]: Let's nuke it from everytwhere if it's not too much work.

File: test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+using Microsoft.AspNetCore.Http;
+using NUnit.Framework;
+
+namespace AutoRest.TestServer.Tests.Infrastructure
+{
+    [TestFixture(TestServerVersion.V1)]
+    public class TestServerLowLevelTestBase
+    {
+        private readonly TestServerVersion _version;
+        internal AzureKeyCredential Key = new AzureKeyCredential("NOT-A-VALID-KEY");
```
Comment (AlexanderSher) [inline test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:21]: Why is it internal?

Comment (chamons) [reply test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:21]: No good reason. Should be protected.

File: test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+        public Task Test(Func<Uri, AzureKeyCredential, Task> test, bool ignoreScenario = false)
+        {
+            return Test(GetScenarioName(), test, ignoreScenario);
+        }
+
+        private async Task Test(string scenario, Func<Uri, AzureKeyCredential, Task> test, bool ignoreScenario = false)
+        {
+            var scenarioParameter = ignoreScenario ? new string[0] : new[] {scenario};
+            var server = TestServerSession.Start(scenario, _version, false, scenarioParameter);
```
Comment (AlexanderSher) [inline test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:63]: `using`?

File: test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+using Azure;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Microsoft.AspNetCore.Http;
+using NUnit.Framework;
+
+namespace AutoRest.TestServer.Tests.Infrastructure
+{
+    [TestFixture(TestServerVersion.V1)]
+    public class TestServerLowLevelTestBase
```
Comment (AlexanderSher) [inline test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:18]: `abstract`?

File: test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+        private Task TestStatus(string scenario, Func<Uri, AzureKeyCredential, Task<Response>> test, bool ignoreScenario = false) => Test(scenario, async (host, key) =>
+        {
+            var response = await test(host, key);
+            Assert.That(response.Status, Is.EqualTo(200).Or.EqualTo(201).Or.EqualTo(202).Or.EqualTo(204), "Unexpected response " + response.ReasonPhrase);
+        }, ignoreScenario);
+
+        public Task Test(Action<Uri, AzureKeyCredential> test, bool ignoreScenario = false)
+        {
+            return Test(GetScenarioName(), (host, key) =>
```
Comment (AlexanderSher) [inline test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:48]:
```cs
public Task Test(Action<Uri, AzureKeyCredential> test, bool ignoreScenario = false) 
    => Test(GetScenarioName(), async (host, key) => test(host, key), ignoreScenario);
```

File: test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs (hunk 5 above)
Comment (chamons) [reply test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:61]: @pakrym - Swapping to async using will change the semantics here somewhat. 

I believe https://gist.github.com/chamons/fcceac847d6f03ed9edef8cb9b680371 is the correct transform. Thoughts?

Comment (chamons) [reply test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:58]: Talked in meeting. Decided to not do this.

**Review by AlexanderSher — APPROVED** [review]

**Comment by chamons** [comment]
Bah - build failing due to launch settings conflict.

**Pull request merged**